crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2372,11 +2372,47 @@ impl<'db> TypeInferenceBuilder<'db> {
             ctx: _,
         } = subscript;
 
-        self.infer_expression(slice);
-        self.infer_expression(value);
+        let value_ty = self.infer_expression(value);
+        let slice_ty = self.infer_expression(slice);
 
-        // TODO actual subscript support
-        Type::Unknown
+        match (value_ty, slice_ty) {
+            // TODO handle variable-length tuples
+            (Type::Tuple(ty), Type::IntLiteral(index)) => {
+                if let Some(ty) = ty.elements(self.db).get(index as usize) {
+                    *ty
+                } else {
+                    Type::Unknown
+                }
+            }
+            (Type::Tuple(ty), Type::BooleanLiteral(bool)) => {
+                if let Some(ty) = ty.elements(self.db).get(usize::from(bool)) {
+                    *ty
+                } else {
+                    Type::Unknown
+                }
+            }
+            (Type::StringLiteral(literal), Type::IntLiteral(index)) => {
+                if let Some(ch) = literal.value(self.db).chars().nth(index as usize) {
+                    Type::StringLiteral(StringLiteralType::new(
+                        self.db,
+                        ch.to_string().into_boxed_str(),
+                    ))
+                } else {
+                    Type::Unknown
+                }
+            }
```

Comment:
You need to add calls to `self.add_diagnostic()` on all these `else {Type::Unknown}` branches (or add a helper method, similar to our `not_iterable_diagnostic()` helper, and call that from each branch)

Reply:
It seems like that shouldn't be "necessary" for the PR to merge given that we have fallbacks to `Type::Unknown` everywhere without diagnostics (`infer_import_definition` would be a good example), but I'll add it here.

Reply:
> `infer_import_definition` would be a good example

We emit a diagnostic here in that method when we can't resolve the import, no? https://github.com/astral-sh/ruff/blob/5f4b2823277df38fc222e9d8096b805f9cc34c74/crates/red_knot_python_semantic/src/types/infer.rs#L1314

Reply:
But yeah, we can defer adding the calls to `add_diagnostic` if you want! But I'd definitely like us to at least add `TODO` comments for these branches if we are going to defer it. Inferring `Unknown` is correct here, but we shouldn't do it silently without emitting a diagnostic to the user informing them about the error, and I think there's a strong chance we'll forget about this otherwise

Reply:
(I was referring to:

```rust
tracing::debug!("Failed to resolve import due to invalid syntax");
Type::Unknown
```
)